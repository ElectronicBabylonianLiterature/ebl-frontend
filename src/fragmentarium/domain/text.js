// @flow
import { immerable } from 'immer'
import Lemmatization, {
  LemmatizationToken
} from 'fragmentarium/domain/Lemmatization'
import type { UniqueLemma } from 'fragmentarium/domain/Lemmatization'

type Word = {|
  +type: 'Word',
  +value: string,
  +uniqueLemma: $ReadOnlyArray<string>,
  +normalized: boolean,
  +language: string,
  +lemmatizable: boolean,
  +erasure: string,
  +alignment?: number
|}

type LoneDeterminative = {|
  +type: 'LoneDeterminative',
  +value: string,
  +uniqueLemma: $ReadOnlyArray<string>,
  +normalized: boolean,
  +language: string,
  +lemmatizable: boolean,
  +partial: [boolean, boolean],
  +erasure: string,
  +alignment?: number
|}

type Shift = {|
  +type: 'LanguageShift',
  +value: string,
  +normalized: boolean,
  +language: string,
  +lemmatizable?: false
|}

type Erasure = {|
  +type: 'Erasure',
  +value: string,
  +side: string,
  +lemmatizable?: false
|}

export type Token =
  | {|
      +type: string,
      +value: string,
      +lemmatizable?: false
    |}
  | Word
  | LoneDeterminative
  | Shift
  | Erasure

export type Line = {|
  +type: 'ControlLine' | 'EmptyLine' | 'TextLine',
  +prefix: string,
  +content: $ReadOnlyArray<Token>
|}

export class Text {
  +lines: $ReadOnlyArray<Line>

  constructor({ lines }: { lines: $ReadOnlyArray<Line> }) {
    this.lines = lines
  }

  createLemmatization(
    lemmas: { [string]: UniqueLemma },
    suggestions: { [string]: $ReadOnlyArray<UniqueLemma> }
  ): Lemmatization {
    return new Lemmatization(
      this.lines.map(line => line.prefix),
      this.lines
        .map(line => line.content)
        .map(tokens =>
          tokens.map(token =>
            token.lemmatizable
              ? new LemmatizationToken(
                  token.value,
                  true,
                  (token.uniqueLemma || []).map(id => lemmas[id]),
                  suggestions[token.value]
                )
              : new LemmatizationToken(token.value, false)
          )
        )
    )
  }
}
Text[immerable] = true
