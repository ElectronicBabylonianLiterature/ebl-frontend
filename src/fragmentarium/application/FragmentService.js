// @flow
import _ from 'lodash'
import Promise from 'bluebird'
import Lemma from 'fragmentarium/domain/Lemma'
import Lemmatization from 'fragmentarium/domain/Lemmatization'
import { Folio, Fragment } from 'fragmentarium/domain/fragment'
import Reference, { createReference } from 'bibliography/domain/Reference'
import { Text } from 'fragmentarium/domain/text'
import type { Token } from 'fragmentarium/domain/text'

export type CdliInfo = {|
  photoUrl: ?string,
  lineArtUrl: ?string,
  detailLineArtUrl: ?string
|}

export interface ImageRepository {
  find(string): Blob;
  findFolio(Folio): Blob;
  findPhoto(string): Blob;
}

export interface FragmentRepository {
  statistics(): Promise<any>;
  find(string): Promise<Fragment>;
  updateTransliteration(string, string, string): Promise<Fragment>;
  updateLemmatization(string, Lemmatization): Promise<Fragment>;
  updateReferences(string, $ReadOnlyArray<Reference>): Promise<Fragment>;
  folioPager(Folio, string): Promise<any>;
  findLemmas(string): Promise<any>;
  fetchCdliInfo(string): Promise<CdliInfo>;
}

class FragmentService {
  #fragmentRepository
  #imageRepository
  #wordRepository
  #bibliographyService

  constructor(
    fragmentRepository: FragmentRepository,
    imageRepository: ImageRepository,
    wordRepository: any,
    bibliographyService: any
  ) {
    this.#fragmentRepository = fragmentRepository
    this.#imageRepository = imageRepository
    this.#wordRepository = wordRepository
    this.#bibliographyService = bibliographyService
  }

  statistics() {
    return this.#fragmentRepository.statistics()
  }

  find(number: string) {
    return this.#fragmentRepository
      .find(number)
      .then(fragment =>
        this.hydrateReferences(fragment.references).then(hydrated =>
          fragment.setReferences(hydrated)
        )
      )
  }

  updateTransliteration(
    number: string,
    transliteration: string,
    notes: string
  ) {
    return this.#fragmentRepository.updateTransliteration(
      number,
      transliteration,
      notes
    )
  }

  updateLemmatization(number: string, lemmatization: Lemmatization) {
    return this.#fragmentRepository.updateLemmatization(number, lemmatization)
  }

  updateReferences(number: string, references: $ReadOnlyArray<Reference>) {
    return this.#fragmentRepository.updateReferences(number, references)
  }

  findFolio(folio: Folio) {
    return this.#imageRepository.findFolio(folio)
  }

  findImage(fileName: string) {
    return this.#imageRepository.find(fileName)
  }

  findPhoto(fragment: Fragment) {
    return fragment.hasPhoto
      ? this.#imageRepository.findPhoto(fragment.number)
      : Promise.resolve(null)
  }

  folioPager(folio: Folio, fragmentNumber: string) {
    return this.#fragmentRepository.folioPager(folio, fragmentNumber)
  }

  searchLemma(lemma: string) {
    return _.isEmpty(lemma)
      ? Promise.resolve([])
      : this.#wordRepository.searchLemma(lemma)
  }

  searchBibliography(query: string) {
    return this.#bibliographyService.search(query)
  }

  fetchCdliInfo(fragment: Fragment) {
    return fragment.cdliNumber
      ? this.#fragmentRepository.fetchCdliInfo(fragment.cdliNumber)
      : Promise.resolve({
          photoUrl: null,
          lineArtUrl: null,
          detailLineArtUrl: null
        })
  }

  createLemmatization(text: Text) {
    return Promise.all([this._fetchLemmas(text), this._fetchSuggestions(text)])
      .then(([lemmaData, suggestionsData]) => [
        _.keyBy(lemmaData, 'value'),
        _.fromPairs(suggestionsData)
      ])
      .then(([lemmas, suggestions]) =>
        text.createLemmatization(lemmas, suggestions)
      )
  }

  _fetchLemmas(text: Text) {
    return Promise.all<$ReadOnlyArray<Lemma>>(
      mapText(
        text,
        line => line.map(token => token.uniqueLemma || []),
        uniqueLemma =>
          this.#wordRepository.find(uniqueLemma).then(word => new Lemma(word))
      )
    )
  }

  _fetchSuggestions(
    text: Text
  ): Promise<$ReadOnlyArray<[string, $ReadOnlyArray<Lemma>]>> {
    return Promise.mapSeries<string, [string, $ReadOnlyArray<Lemma>], any>(
      mapLines(text, line =>
        line.filter(token => token.lemmatizable).map(token => token.value)
      ),
      (value: string): Promise<[string, $ReadOnlyArray<Lemma>]> =>
        this.#fragmentRepository
          .findLemmas(value)
          .then(lemmas => [
            value,
            lemmas.map(complexLemma =>
              complexLemma.map(word => new Lemma(word))
            )
          ])
    )
  }

  hydrateReferences(references: $ReadOnlyArray<{}>) {
    const hydrate: ({}) => Promise<Reference> = reference =>
      createReference(reference, this.#bibliographyService)
    return Promise.all<Reference>(references.map(hydrate))
  }
}

function mapText<T, U>(
  text: Text,
  mapLine: ($ReadOnlyArray<Token>) => ?T,
  mapToken: T => U
): $ReadOnlyArray<U> {
  return mapLines(text, mapLine).map(mapToken)
}

function mapLines<T>(
  text: Text,
  mapLine: ($ReadOnlyArray<Token>) => $ReadOnlyArray<?T> | ?T
): $ReadOnlyArray<T> {
  return _(text.lines)
    .flatMapDeep(({ content }) => mapLine(content))
    .reject(_.isNil)
    .uniq()
    .sort()
    .value()
}

export default FragmentService
