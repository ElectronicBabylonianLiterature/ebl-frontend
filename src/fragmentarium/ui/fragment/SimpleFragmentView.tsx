import React from 'react'
import { Fragment } from 'fragmentarium/domain/fragment'
import withData from 'http/withData'
import FragmentService from 'fragmentarium/application/FragmentService'
import { Session } from 'auth/Session'
import { Transliteration } from 'transliteration/ui/Transliteration'
import './SimpleFragmentView.sass'
import { useHistory } from 'react-router-dom'
import { parse } from 'query-string'

function getLanguageUrlParam(query: string): string | undefined {
  const lang = parse(query).lang
  return typeof lang === 'string' ? lang : undefined
}

function SimpleView({
  session,
  fragment,
}: {
  session: Session
  fragment: Fragment
}): JSX.Element {
  const { location } = useHistory()
  return session.isAllowedToReadFragments() ? (
    <>
      <h1 className={'SimpleFragmentView__title'}>{fragment.number}</h1>
      <Transliteration
        text={fragment.text}
        language={getLanguageUrlParam(location.search)}
      />
    </>
  ) : (
    <>{"You don't have permissions to view this fragment."}</>
  )
}

const SimpleFragmentView = withData<
  {
    fragmentService: FragmentService
    session: Session
  },
  { number: string },
  Fragment
>(
  ({ data, ...props }) => <SimpleView fragment={data} {...props} />,
  (props) =>
    props.fragmentService
      .find(props.number)
      .then((fragment) => fragment.filterFolios(props.session)),
  {
    watch: (props) => [props.number],
  }
)
export default SimpleFragmentView
