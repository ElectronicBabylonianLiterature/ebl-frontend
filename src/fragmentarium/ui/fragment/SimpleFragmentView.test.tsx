import React from 'react'
import { render, act, screen } from '@testing-library/react'
import MemorySession from 'auth/Session'
import FragmentService from 'fragmentarium/application/FragmentService'
import SimpleFragmentView from 'fragmentarium/ui/fragment/SimpleFragmentView'
import { createMemoryHistory, MemoryHistory } from 'history'
import { Router } from 'react-router-dom'
import { fragmentFactory } from 'test-support/fragment-fixtures'
import { Text } from 'transliteration/domain/text'
import textLine, { textLineDto } from 'test-support/lines/text-line'
import {
  arabicTranslationLine,
  englishTranslationLine,
} from 'test-support/lines/translation-lines'
import { TextLine } from 'transliteration/domain/text-line'
import { lineNumberFactory } from 'test-support/linenumber-factory'

jest.mock('fragmentarium/application/FragmentService')
jest.mock('auth/Session')

const MockFragmentService = FragmentService as jest.Mock<
  jest.Mocked<FragmentService>
>
const fragmentServiceMock = new MockFragmentService()
const session = new (MemorySession as jest.Mock<jest.Mocked<MemorySession>>)()
let container: HTMLElement
let history: MemoryHistory

const textLine2 = new TextLine({
  ...textLineDto,
  lineNumber: lineNumberFactory.build({ number: 2 }),
})

const translatedText = new Text({
  lines: [textLine, textLine2, englishTranslationLine, arabicTranslationLine],
})
const translatedFragment = fragmentFactory.build({
  number: 'Translated.Fragment',
  text: translatedText,
})

async function renderSimpleFragmentView(history: MemoryHistory) {
  await act(async () => {
    container = render(
      <Router history={history}>
        <SimpleFragmentView
          fragmentService={fragmentServiceMock}
          number={translatedFragment.number}
          session={session}
        />
      </Router>
    ).container
  })
}

beforeEach(async () => {
  session.isAllowedToReadFragments.mockReturnValue(true)
  fragmentServiceMock.find.mockResolvedValue(translatedFragment)
})

it('correctly displays the simple fragment view', async () => {
  history = createMemoryHistory()
  await renderSimpleFragmentView(history)

  expect(container).toMatchSnapshot()
})
describe('language url parameter', () => {
  const renderWithLanguage = async (language: string) => {
    history = createMemoryHistory({
      initialEntries: [`/html?lang=${language}`],
    })
    await renderSimpleFragmentView(history)
  }

  test.each([
    ['en', 'English', 'Arabic'],
    ['ar', 'Arabic', 'English'],
  ])(
    'only shows %s when lang=%s is set',
    async (lang, visibleLang, hiddenLang) => {
      await renderWithLanguage(lang)
      expect(screen.getByText(`${visibleLang} translation`)).toBeVisible()
      expect(
        screen.queryByText(`${hiddenLang} translation`)
      ).not.toBeInTheDocument()
    }
  )

  it('shows all languages when lang is not set', async () => {
    history = createMemoryHistory()
    await renderSimpleFragmentView(history)
    expect(screen.getByText('English translation')).toBeVisible()
    expect(screen.getByText('Arabic translation')).toBeVisible()
  })
})
