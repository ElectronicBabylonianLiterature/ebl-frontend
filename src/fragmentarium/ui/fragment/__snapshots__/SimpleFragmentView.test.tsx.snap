// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly displays the simple fragment view 1`] = `
<div>
  <h1
    class="SimpleFragmentView__title"
  >
    Translated.Fragment
  </h1>
  <section
    class="Transliteration"
  >
    <table
      class="Transliteration__lines"
    >
      <tbody>
        <tr
          id="line-1"
        >
          <td
            class="Transliteration__TextLine"
          >
            <a
              class="Transliteration__anchor"
              href="#%201"
              id=" 1"
            >
              <sup>
                (
                1
                )
              </sup>
            </a>
          </td>
          <td
            colspan="1"
          >
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class=""
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class=""
                      >
                        érin
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class=""
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__BrokenAway"
                >
                  <span
                    class=""
                  >
                    [
                  </span>
                </span>
                <span
                  class="Transliteration__PerhapsBrokenAway"
                >
                  <span
                    class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY"
                  >
                    (
                  </span>
                </span>
                <span
                  class="Transliteration__Logogram"
                >
                  <span
                    class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                      >
                        ŠÁ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class="Transliteration__Joiner--BROKEN_AWAY Transliteration__Joiner--PERHAPS_BROKEN_AWAY"
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__Variant"
                >
                  <span
                    class="Transliteration__Reading"
                  >
                    <span
                      class="Transliteration__Reading--BROKEN_AWAY Transliteration__Reading--PERHAPS_BROKEN_AWAY"
                    >
                      <span
                        class="Transliteration__ValueToken"
                      >
                        <span
                          class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                        >
                          šà
                        </span>
                      </span>
                      <sup
                        class="Transliteration__modifier"
                      />
                      <sup
                        class="Transliteration__flag"
                      />
                    </span>
                  </span>
                  /
                  <span
                    class="Transliteration__Logogram"
                  >
                    <span
                      class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                    >
                      <span
                        class="Transliteration__ValueToken"
                      >
                        <span
                          class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                        >
                          GÌRI
                        </span>
                      </span>
                      <sup
                        class="Transliteration__modifier"
                      />
                      <sup
                        class="Transliteration__flag"
                      />
                    </span>
                  </span>
                </span>
                <span
                  class="Transliteration__PerhapsBrokenAway"
                >
                  <span
                    class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY Transliteration__PerhapsBrokenAway--PERHAPS_BROKEN_AWAY"
                  >
                    )
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class=""
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class=""
                      >
                        kur
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class="Transliteration__Word--BROKEN_AWAY"
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class="Transliteration__Reading--BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY"
                      >
                        saḫ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class="Transliteration__Joiner--BROKEN_AWAY"
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__Logogram"
                >
                  <span
                    class="Transliteration__Logogram--BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY"
                      >
                        SAḪ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__UnknownNumberOfSigns Transliteration__UnknownNumberOfSigns--AKKADIAN"
            >
              …
            </span>
            <span
              class="Transliteration__BrokenAway Transliteration__BrokenAway--AKKADIAN"
            >
              <span
                class="Transliteration__BrokenAway--BROKEN_AWAY"
              >
                ]
              </span>
            </span>
          </td>
          <td>
             
          </td>
        </tr>
        <tr
          id="line-2"
        >
          <td
            class="Transliteration__TranslationLine Transliteration__TranslationLine--extent"
          >
            en
            :
          </td>
          <td
            class="Transliteration__TranslationLine"
            colspan="1"
          >
            <div>
              <span>
                English translation
              </span>
            </div>
          </td>
          <td>
             
          </td>
        </tr>
        <tr
          id="line-3"
        >
          <td
            class="Transliteration__TranslationLine Transliteration__TranslationLine--extent"
          >
            ar
            :
          </td>
          <td
            class="Transliteration__TranslationLine"
            colspan="1"
          >
            <div>
              <span>
                Arabic translation
              </span>
            </div>
          </td>
          <td>
             
          </td>
        </tr>
        <tr
          id="line-4"
        >
          <td
            class="Transliteration__TextLine"
          >
            <a
              class="Transliteration__anchor"
              href="#%202"
              id=" 2"
            >
              <sup>
                (
                2
                )
              </sup>
            </a>
          </td>
          <td
            colspan="1"
          >
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class=""
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class=""
                      >
                        érin
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class=""
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__BrokenAway"
                >
                  <span
                    class=""
                  >
                    [
                  </span>
                </span>
                <span
                  class="Transliteration__PerhapsBrokenAway"
                >
                  <span
                    class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY"
                  >
                    (
                  </span>
                </span>
                <span
                  class="Transliteration__Logogram"
                >
                  <span
                    class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                      >
                        ŠÁ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class="Transliteration__Joiner--BROKEN_AWAY Transliteration__Joiner--PERHAPS_BROKEN_AWAY"
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__Variant"
                >
                  <span
                    class="Transliteration__Reading"
                  >
                    <span
                      class="Transliteration__Reading--BROKEN_AWAY Transliteration__Reading--PERHAPS_BROKEN_AWAY"
                    >
                      <span
                        class="Transliteration__ValueToken"
                      >
                        <span
                          class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                        >
                          šà
                        </span>
                      </span>
                      <sup
                        class="Transliteration__modifier"
                      />
                      <sup
                        class="Transliteration__flag"
                      />
                    </span>
                  </span>
                  /
                  <span
                    class="Transliteration__Logogram"
                  >
                    <span
                      class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                    >
                      <span
                        class="Transliteration__ValueToken"
                      >
                        <span
                          class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                        >
                          GÌRI
                        </span>
                      </span>
                      <sup
                        class="Transliteration__modifier"
                      />
                      <sup
                        class="Transliteration__flag"
                      />
                    </span>
                  </span>
                </span>
                <span
                  class="Transliteration__PerhapsBrokenAway"
                >
                  <span
                    class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY Transliteration__PerhapsBrokenAway--PERHAPS_BROKEN_AWAY"
                  >
                    )
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class=""
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class=""
                      >
                        kur
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class="Transliteration__Word--BROKEN_AWAY"
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class="Transliteration__Reading--BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY"
                      >
                        saḫ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class="Transliteration__Joiner--BROKEN_AWAY"
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__Logogram"
                >
                  <span
                    class="Transliteration__Logogram--BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY"
                      >
                        SAḪ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__UnknownNumberOfSigns Transliteration__UnknownNumberOfSigns--AKKADIAN"
            >
              …
            </span>
            <span
              class="Transliteration__BrokenAway Transliteration__BrokenAway--AKKADIAN"
            >
              <span
                class="Transliteration__BrokenAway--BROKEN_AWAY"
              >
                ]
              </span>
            </span>
          </td>
          <td>
             
          </td>
        </tr>
        <tr
          id="line-5"
        >
          <td
            class="Transliteration__TranslationLine Transliteration__TranslationLine--extent"
          >
            en
             
            (
            
            3
            )
            :
          </td>
          <td
            class="Transliteration__TranslationLine"
            colspan="1"
          >
            <div>
              <span>
                English translation over two lines
              </span>
            </div>
          </td>
          <td>
             
          </td>
        </tr>
        <tr
          id="line-6"
        >
          <td
            class="Transliteration__TextLine"
          >
            <a
              class="Transliteration__anchor"
              href="#%203"
              id=" 3"
            >
              <sup>
                (
                3
                )
              </sup>
            </a>
          </td>
          <td
            colspan="1"
          >
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class=""
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class=""
                      >
                        érin
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class=""
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__BrokenAway"
                >
                  <span
                    class=""
                  >
                    [
                  </span>
                </span>
                <span
                  class="Transliteration__PerhapsBrokenAway"
                >
                  <span
                    class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY"
                  >
                    (
                  </span>
                </span>
                <span
                  class="Transliteration__Logogram"
                >
                  <span
                    class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                      >
                        ŠÁ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class="Transliteration__Joiner--BROKEN_AWAY Transliteration__Joiner--PERHAPS_BROKEN_AWAY"
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__Variant"
                >
                  <span
                    class="Transliteration__Reading"
                  >
                    <span
                      class="Transliteration__Reading--BROKEN_AWAY Transliteration__Reading--PERHAPS_BROKEN_AWAY"
                    >
                      <span
                        class="Transliteration__ValueToken"
                      >
                        <span
                          class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                        >
                          šà
                        </span>
                      </span>
                      <sup
                        class="Transliteration__modifier"
                      />
                      <sup
                        class="Transliteration__flag"
                      />
                    </span>
                  </span>
                  /
                  <span
                    class="Transliteration__Logogram"
                  >
                    <span
                      class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                    >
                      <span
                        class="Transliteration__ValueToken"
                      >
                        <span
                          class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                        >
                          GÌRI
                        </span>
                      </span>
                      <sup
                        class="Transliteration__modifier"
                      />
                      <sup
                        class="Transliteration__flag"
                      />
                    </span>
                  </span>
                </span>
                <span
                  class="Transliteration__PerhapsBrokenAway"
                >
                  <span
                    class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY Transliteration__PerhapsBrokenAway--PERHAPS_BROKEN_AWAY"
                  >
                    )
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class=""
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class=""
                      >
                        kur
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__Word Transliteration__Word--AKKADIAN"
            >
              <span
                class="Transliteration__Word--BROKEN_AWAY"
              >
                <span
                  class="Transliteration__Reading"
                >
                  <span
                    class="Transliteration__Reading--BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY"
                      >
                        saḫ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
                <span
                  class="Transliteration__Joiner"
                >
                  <span
                    class="Transliteration__Joiner--BROKEN_AWAY"
                  >
                    -
                  </span>
                </span>
                <span
                  class="Transliteration__Logogram"
                >
                  <span
                    class="Transliteration__Logogram--BROKEN_AWAY"
                  >
                    <span
                      class="Transliteration__ValueToken"
                    >
                      <span
                        class="Transliteration__ValueToken--BROKEN_AWAY"
                      >
                        SAḪ
                      </span>
                    </span>
                    <sup
                      class="Transliteration__modifier"
                    />
                    <sup
                      class="Transliteration__flag"
                    />
                  </span>
                </span>
              </span>
            </span>
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__UnknownNumberOfSigns Transliteration__UnknownNumberOfSigns--AKKADIAN"
            >
              …
            </span>
            <span
              class="Transliteration__BrokenAway Transliteration__BrokenAway--AKKADIAN"
            >
              <span
                class="Transliteration__BrokenAway--BROKEN_AWAY"
              >
                ]
              </span>
            </span>
          </td>
          <td>
             
          </td>
        </tr>
      </tbody>
    </table>
    <ol
      class="Transliteration__notes"
    />
  </section>
</div>
`;
